// mdb-angular-ui-kit/esm2022/scrollspy/scrollspy.module.mjs
import { NgModule } from '@angular/core';
import { MdbScrollspyDirective } from './scrollspy.directive';
import { MdbScrollspyLinkDirective } from './scrollspy-link.directive';
import { MdbScrollspyElementDirective } from './scrollspy-element.directive';
import { MdbScrollspyService } from './scrollspy.service';
import { MdbScrollspyWindowDirective } from './scrollspy-window.directive';
import * as i0 from "@angular/core";
export class MdbScrollspyModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.0.7", ngImport: i0, type: MdbScrollspyModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.0.7", ngImport: i0, type: MdbScrollspyModule, declarations: [MdbScrollspyDirective,
            MdbScrollspyLinkDirective,
            MdbScrollspyElementDirective,
            MdbScrollspyWindowDirective], exports: [MdbScrollspyDirective,
            MdbScrollspyLinkDirective,
            MdbScrollspyElementDirective,
            MdbScrollspyWindowDirective] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.0.7", ngImport: i0, type: MdbScrollspyModule, providers: [MdbScrollspyService] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.0.7", ngImport: i0, type: MdbScrollspyModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        MdbScrollspyDirective,
                        MdbScrollspyLinkDirective,
                        MdbScrollspyElementDirective,
                        MdbScrollspyWindowDirective,
                    ],
                    exports: [
                        MdbScrollspyDirective,
                        MdbScrollspyLinkDirective,
                        MdbScrollspyElementDirective,
                        MdbScrollspyWindowDirective,
                    ],
                    providers: [MdbScrollspyService],
                }]
        }] });
