// mdb-angular-ui-kit/esm2022/carousel/carousel.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MdbCarouselComponent } from './carousel.component';
import { MdbCarouselItemComponent } from './carousel-item.component';
import * as i0 from "@angular/core";
export class MdbCarouselModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.0.7", ngImport: i0, type: MdbCarouselModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.0.7", ngImport: i0, type: MdbCarouselModule, declarations: [MdbCarouselComponent, MdbCarouselItemComponent], imports: [CommonModule], exports: [MdbCarouselComponent, MdbCarouselItemComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.0.7", ngImport: i0, type: MdbCarouselModule, imports: [CommonModule] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.0.7", ngImport: i0, type: MdbCarouselModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [MdbCarouselComponent, MdbCarouselItemComponent],
                    exports: [MdbCarouselComponent, MdbCarouselItemComponent],
                    imports: [CommonModule],
                }]
        }] });
