// mdb-angular-ui-kit/esm2022/radio/radio.module.mjs
import { MdbRadioDirective } from './radio-button.directive';
import { MdbRadioGroupDirective } from './radio-group.directive';
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import * as i0 from "@angular/core";
export class MdbRadioModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.0.7", ngImport: i0, type: MdbRadioModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.0.7", ngImport: i0, type: MdbRadioModule, declarations: [MdbRadioDirective, MdbRadioGroupDirective], imports: [CommonModule, FormsModule], exports: [MdbRadioDirective, MdbRadioGroupDirective] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.0.7", ngImport: i0, type: MdbRadioModule, imports: [CommonModule, FormsModule] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.0.7", ngImport: i0, type: MdbRadioModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [MdbRadioDirective, MdbRadioGroupDirective],
                    exports: [MdbRadioDirective, MdbRadioGroupDirective],
                    imports: [CommonModule, FormsModule],
                }]
        }] });
