// mdb-angular-ui-kit/esm2022/accordion/accordion.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MdbAccordionComponent } from './accordion.component';
import { MdbAccordionItemComponent } from './accordion-item.component';
import { MdbAccordionItemHeaderDirective } from './accordion-item-header.directive';
import { MdbAccordionItemBodyDirective } from './accordion-item-content.directive';
import { MdbCollapseModule } from 'mdb-angular-ui-kit/collapse';
import * as i0 from "@angular/core";
export class MdbAccordionModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.0.7", ngImport: i0, type: MdbAccordionModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.0.7", ngImport: i0, type: MdbAccordionModule, declarations: [MdbAccordionComponent,
            MdbAccordionItemComponent,
            MdbAccordionItemHeaderDirective,
            MdbAccordionItemBodyDirective], imports: [CommonModule, MdbCollapseModule], exports: [MdbAccordionComponent,
            MdbAccordionItemComponent,
            MdbAccordionItemHeaderDirective,
            MdbAccordionItemBodyDirective] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.0.7", ngImport: i0, type: MdbAccordionModule, imports: [CommonModule, MdbCollapseModule] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.0.7", ngImport: i0, type: MdbAccordionModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        MdbAccordionComponent,
                        MdbAccordionItemComponent,
                        MdbAccordionItemHeaderDirective,
                        MdbAccordionItemBodyDirective,
                    ],
                    imports: [CommonModule, MdbCollapseModule],
                    exports: [
                        MdbAccordionComponent,
                        MdbAccordionItemComponent,
                        MdbAccordionItemHeaderDirective,
                        MdbAccordionItemBodyDirective,
                    ],
                }]
        }] });
