// mdb-angular-ui-kit/esm2022/tabs/tabs.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MdbTabComponent } from './tab.component';
import { MdbTabsComponent } from './tabs.component';
import { PortalModule } from '@angular/cdk/portal';
import { MdbTabContentDirective } from './tab-content.directive';
import { MdbTabPortalOutlet } from './tab-outlet.directive';
import { MdbTabTitleDirective } from './tab-title.directive';
import * as i0 from "@angular/core";
export class MdbTabsModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.0.7", ngImport: i0, type: MdbTabsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.0.7", ngImport: i0, type: MdbTabsModule, declarations: [MdbTabComponent,
            MdbTabContentDirective,
            MdbTabTitleDirective,
            MdbTabPortalOutlet,
            MdbTabsComponent], imports: [CommonModule, PortalModule], exports: [MdbTabComponent,
            MdbTabContentDirective,
            MdbTabTitleDirective,
            MdbTabPortalOutlet,
            MdbTabsComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.0.7", ngImport: i0, type: MdbTabsModule, imports: [CommonModule, PortalModule] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.0.7", ngImport: i0, type: MdbTabsModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        MdbTabComponent,
                        MdbTabContentDirective,
                        MdbTabTitleDirective,
                        MdbTabPortalOutlet,
                        MdbTabsComponent,
                    ],
                    imports: [CommonModule, PortalModule],
                    exports: [
                        MdbTabComponent,
                        MdbTabContentDirective,
                        MdbTabTitleDirective,
                        MdbTabPortalOutlet,
                        MdbTabsComponent,
                    ],
                }]
        }] });
